File: reference/TokenHandlerMinimal.java
import java.net.URI;
import java.net.http.*;
import java.time.Instant;
import java.util.function.Supplier;

/**
 * Weniger ist mehr — Single-Flight-Token-Cache OHNE MS-SDK, ~12 Zeilen Kern.
 * N parallele Aufrufer  ->  GENAU EIN Refresh (Double-Checked Locking).
 *
 * Der „Klick" sind drei Schritte:
 *   1. Gültig?           -> volatile-Read, lock-frei zurück (der 99-%-Fall).
 *   2. Nein -> synchronized, NOCHMAL prüfen (jemand war schneller).
 *   3. Immer noch nichts -> selbst ziehen. Weil das im synchronized-Block
 *      passiert, zieht von 100 Threads nur EINER.
 *
 * Das Anti-Pattern, das die Prod-Race erzeugt, wäre:  if (expired) fetch();
 * OHNE Lock -> 100 Threads ziehen gleichzeitig 100 Tokens und überschreiben
 * sich den Cache ("Bearer null"-Fenster).
 */
public final class TokenHandlerMinimal {

    public record Token(String value, Instant expiresAt) {}

    private final Supplier<Token> fetch;   // () -> Token  (der eigentliche Bezug, s. unten)
    private final long skewSeconds;        // so viele Sekunden VOR Ablauf schon erneuern
    private volatile Token current;        // volatile => Fast Path ohne Lock

    public TokenHandlerMinimal(Supplier<Token> fetch, long skewSeconds) {
        this.fetch = fetch;
        this.skewSeconds = skewSeconds;
    }

    public String get() {
        Token t = current;                                            // (1) lock-frei
        if (t != null && Instant.now().isBefore(t.expiresAt().minusSeconds(skewSeconds)))
            return t.value();
        synchronized (this) {                                         // (2)+(3) nur EINER zieht
            t = current;
            if (t != null && Instant.now().isBefore(t.expiresAt().minusSeconds(skewSeconds)))
                return t.value();                                     //     jemand war schneller
            current = fetch.get();                                    //     genau ein Refresh
            return current.value();
        }
    }

    // -----------------------------------------------------------------------
    // „Ohne SDK" heißt auch: den Token-Bezug selbst schreiben. Illustrativer
    // hand-gerollter Client-Credentials-POST (java.net.http). In echt: JSON
    // sauber parsen (Jackson) statt der naiven Extraktion hier.
    // -----------------------------------------------------------------------
    static Token fetchViaHttp(String tenantId, String clientId, String secret, String scope) {
        try {
            String body = "client_id=" + clientId
                        + "&client_secret=" + secret
                        + "&scope=" + java.net.URLEncoder.encode(scope, "UTF-8")
                        + "&grant_type=client_credentials";
            HttpResponse<String> r = HttpClient.newHttpClient().send(
                HttpRequest.newBuilder(URI.create(
                    "https://login.microsoftonline.com/" + tenantId + "/oauth2/v2.0/token"))
                    .header("Content-Type", "application/x-www-form-urlencoded")
                    .POST(HttpRequest.BodyPublishers.ofString(body)).build(),
                HttpResponse.BodyHandlers.ofString());
            if (r.statusCode() != 200) throw new RuntimeException("token endpoint " + r.statusCode());
            String access = between(r.body(), "\"access_token\":\"", "\"");
            long ttl = Long.parseLong(between(r.body(), "\"expires_in\":", ",").trim());
            return new Token(access, Instant.now().plusSeconds(ttl));
        } catch (Exception e) { throw new RuntimeException(e); }
    }
    private static String between(String s, String a, String b) {
        int i = s.indexOf(a) + a.length();
        return s.substring(i, s.indexOf(b, i));
    }
}

File: reference/token_handler_minimal.py
"""
Weniger ist mehr — der KERN eines Single-Flight-Token-Caches in ~15 Zeilen.
Das ist der „Klick": N parallele Aufrufer  ->  GENAU EIN Refresh.

Die ganze Idee sind drei Schritte:

  1. Gültiges Token da?  ->  ohne Lock zurück.        (der Normalfall, ~99 %)
  2. Nein -> Lock nehmen und NOCHMAL prüfen.          (Double-Check!)
     Jemand war in der Zwischenzeit schneller und hat schon erneuert
     -> dessen frisches Token einfach mitnehmen.
  3. Immer noch nichts -> selbst ziehen.
     Weil das UNTER dem Lock passiert, zieht von 100 Threads nur EINER;
     die anderen bekommen in Schritt 2 das frische Token.

Das ist „Double-Checked Locking". Mehr braucht der Kern nicht.
Budget, Retry, Refresh-Ahead und Circuit-Breaker sind Härtung obendrauf
(siehe authoritative_token_handler.py) — aber DAS hier ist der Aha-Moment.
"""
import time
import threading


class TokenHandler:
    def __init__(self, fetch, skew=300):
        self._fetch = fetch          # callable: () -> (token: str, expires_at_epoch: float)
        self._skew = skew            # so viele Sekunden VOR Ablauf schon erneuern
        self._token, self._exp = None, 0.0
        self._lock = threading.Lock()

    def get(self) -> str:
        # (1) FAST PATH — gültig? -> lock-frei zurück. Der 99-%-Fall.
        if self._token and time.time() < self._exp - self._skew:
            return self._token

        # (2)+(3) SLOW PATH — nur EINER zieht, der Rest wartet am Lock.
        with self._lock:
            if self._token and time.time() < self._exp - self._skew:
                return self._token                     # (2) jemand war schneller
            self._token, self._exp = self._fetch()     # (3) genau ein Refresh
            return self._token


# ---------------------------------------------------------------------------
# Warum das reicht (und warum es klickt):
#   - Der Fast Path braucht KEIN Lock -> 10k gleichzeitige Requests mit
#     gültigem Token konkurrieren um nichts.
#   - Der Slow Path ist durch den Lock serialisiert, ABER die zweite Prüfung
#     sorgt dafür, dass trotzdem nur EIN echter Refresh passiert.
#   - Kein `if expired: fetch()` ohne Lock -> genau das wäre die Race, bei der
#     100 Threads gleichzeitig 100 Tokens ziehen und sich gegenseitig den Cache
#     überschreiben ("Bearer None"-Fenster).
# ---------------------------------------------------------------------------
